state/src/storage/cursor.rs

```diff
@@ -20,104 +20,178 @@ use super::{
+    ///
+    /// Returns `true` if the cursor has reached the start of the table
+    pub fn seek_prev(&mut self) -> Result<bool, StorageError> {
+        // Move the cursor back and return early if it is exhausted
+        if self.inner.prev::<CowBuffer, CowBuffer>().map_err(StorageError::TxOp)?.is_none() {
+            return Ok(true);
+        }
 
-        res.map(|(k, v)| Self::deserialize_key_value(&k, &v)).transpose()
+        // Seek backward to the previous valid key
+        let start = self.seek_next_filtered(false /* forward */)?.is_none();
+        Ok(start)
     }
 
     /// Seek to the first key in the table
     pub fn seek_first(&mut self) -> Result<(), StorageError> {
-        let res = self.inner.first::<CowBuffer, CowBuffer>().map_err(StorageError::TxOp)?;
-
-        self.buffer_kv_pair(res)
+        // Position the cursor at the first key then seek forward to the next valid key
+        self.inner.first::<CowBuffer, CowBuffer>().map_err(StorageError::TxOp)?;
+        self.seek_next_filtered(true /* forward */).map(|_| ())
     }
 
     /// Seek to the last key in the table
     pub fn seek_last(&mut self) -> Result<(), StorageError> {
-        let res = self.inner.last::<CowBuffer, CowBuffer>().map_err(StorageError::TxOp)?;
-
-        self.buffer_kv_pair(res)
+        // Position the cursor at the last key then seek backwards to the next valid key
+        self.inner.last::<CowBuffer, CowBuffer>().map_err(StorageError::TxOp)?;
+        self.seek_next_filtered(false /* forward */).map(|_| ())
     }
 
     /// Position the cursor at a specific key
+    ///
+    /// Note that if the key is not present, the cursor is positioned at an
+    /// empty value, the filter will then take effect on any further cursor
+    /// operations
     pub fn seek(&mut self, k: &K) -> Result<(), StorageError> {
-        let k_bytes = serialize_value(&k)?;
-        let res =
-            self.inner.set_key::<CowBuffer, CowBuffer>(&k_bytes).map_err(StorageError::TxOp)?;
+        // Validate the key
+        if let Some(filter) = self.key_filter
+            && !filter(k)
+        {
+            return Err(StorageError::InvalidKey(format!(
+                "Key {k:?} passed to `Cursor::seek` does not pass filter"
+            )));
+        }
 
-        self.buffer_kv_pair(res)
+        let k_bytes = serialize_value(&k)?;
+        self.inner.set_key::<CowBuffer, CowBuffer>(&k_bytes).map_err(StorageError::TxOp).map(|_| ())
     }
 
     /// Position at the first key greater than or equal to the given key
     pub fn seek_geq(&mut self, k: &K) -> Result<(), StorageError> {
+        // Position the cursor at the first key then seek forward to the next valid key
         let k_bytes = serialize_value(k)?;
-        let res =
-            self.inner.set_range::<CowBuffer, CowBuffer>(&k_bytes).map_err(StorageError::TxOp)?;
-
-        self.buffer_kv_pair(res)
+        self.inner.set_range::<CowBuffer, CowBuffer>(&k_bytes).map_err(StorageError::TxOp)?;
```

Comment:
should we check that the first key in the range matches the filter? If it doesn't, it gives us a chance to error early instead of potentially seeking to the end of the table and having the cursor appear empty

Reply:
I prefer it this way, that isn't really an erroneous case, it's just an empty cursor. This matches what would happen if _no_ filter is applied, i.e. if I have $1, \ldots, 9$ in the table and do `cursor.seek_geq("10")` I'll just get an empty cursor, not an error. It doesn't seem sensible to error only if a filter is there.

Plus semantically, empty range queries are completely valid and useful